core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -116,174 +118,179 @@ sealed trait Channel[F[_], A] {
-    bounded(0)
-
-  def bounded[F[_], A](capacity: Int)(implicit F: Concurrent[F]): F[Channel[F, A]] = {
-    case class State(
-        values: List[A],
-        size: Int,
-        waiting: Option[Deferred[F, Unit]],
-        producers: List[(A, Deferred[F, Unit])],
-        closed: Boolean
-    )
-
-    val open = State(List.empty, 0, None, List.empty, closed = false)
-
-    def empty(isClosed: Boolean): State =
-      if (isClosed) State(List.empty, 0, None, List.empty, closed = true)
-      else open
-
-    (F.ref(open), F.deferred[Unit]).mapN { (state, closedGate) =>
-      new Channel[F, A] {
-
-        def sendAll: Pipe[F, A, Nothing] = { in =>
-          (in ++ Stream.exec(close.void))
-            .evalMap(send)
-            .takeWhile(_.isRight)
-            .drain
-        }
-
-        def send(a: A) =
-          F.deferred[Unit].flatMap { producer =>
-            F.uncancelable { poll =>
-              state.modify {
-                case s @ State(_, _, _, _, closed @ true) =>
-                  (s, Channel.closed[Unit].pure[F])
-
-                case State(values, size, waiting, producers, closed @ false) =>
-                  if (size < capacity)
-                    (
-                      State(a :: values, size + 1, None, producers, false),
-                      notifyStream(waiting).as(rightUnit)
-                    )
+    Queue.synchronous[F, AnyRef].flatMap(impl(_))
+
+  def bounded[F[_], A](capacity: Int)(implicit F: Concurrent[F]): F[Channel[F, A]] =
+    Queue.bounded[F, AnyRef](capacity).flatMap(impl(_))
+
+  // used as a marker to wake up q.take when the channel is closed
+  private[this] val Sentinel = new AnyRef
+
+  private[this] val LeftClosed: Either[Channel.Closed, Unit] = Left(Channel.Closed)
+  private[this] val RightUnit: Either[Channel.Closed, Unit] = Right(())
+
+  // technically this should be A | Sentinel.type
+  // the queue will consist of exclusively As until we shut down, when there will be one Sentinel
+  private[this] def impl[F[_]: Concurrent, A](q: Queue[F, AnyRef]): F[Channel[F, A]] =
+    (Concurrent[F].ref(0), Concurrent[F].ref(false), Concurrent[F].deferred[Unit]).mapN {
+      (leasesR, closedR, closedLatch) =>
+        new Channel[F, A] {
+
+          private[this] val LeftClosedF = LeftClosed.pure[F]
+          private[this] val FalseF = false.pure[F]
+
+          // might be interesting to try to optimize this more, but it needs support from CE
+          val sendAll: Pipe[F, A, Nothing] =
+            _.evalMapChunk(send(_))
+              .takeWhile(_.isRight)
+              .onComplete(Stream.exec(close.void))
+              .drain
+
+          // setting the flag means we won't accept any more sends
+          val close: F[Either[Channel.Closed, Unit]] =
+            closedR.getAndSet(true).flatMap { b =>
+              if (b) {
+                LeftClosedF
+              } else {
+                leasesR.get.flatMap { leases =>
+                  if (leases <= 0)
```

Comment:
When can this be negative?

Reply:
Never. I just habitually defensively code my bounds checks. I'm happy to shift this to a `==` if you like